{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/gdakshareddybt23/Bioinformatics/blob/main/De_Novo_Genome_Assembly.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "GZ6eHTtWv2Z5",
        "outputId": "bf040333-9550-4e5b-eca9-2a569c0c6a15"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "fastqc is already the newest version (0.11.9+dfsg-5).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "velvet is already the newest version (1.2.10+dfsg1-7).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.\n",
            "--2026-01-06 05:22:41--  https://zenodo.org/record/582600/files/mutant_R1.fastq\n",
            "Resolving zenodo.org (zenodo.org)... 137.138.52.235, 188.185.43.153, 188.185.48.75, ...\n",
            "Connecting to zenodo.org (zenodo.org)|137.138.52.235|:443... connected.\n",
            "HTTP request sent, awaiting response... 301 MOVED PERMANENTLY\n",
            "Location: /records/582600/files/mutant_R1.fastq [following]\n",
            "--2026-01-06 05:22:41--  https://zenodo.org/records/582600/files/mutant_R1.fastq\n",
            "Reusing existing connection to zenodo.org:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 4137809 (3.9M) [application/octet-stream]\n",
            "Saving to: ‘mutant_R1.fastq’\n",
            "\n",
            "mutant_R1.fastq     100%[===================>]   3.95M  3.19MB/s    in 1.2s    \n",
            "\n",
            "2026-01-06 05:22:43 (3.19 MB/s) - ‘mutant_R1.fastq’ saved [4137809/4137809]\n",
            "\n",
            "--2026-01-06 05:22:43--  https://zenodo.org/record/582600/files/mutant_R2.fastq\n",
            "Resolving zenodo.org (zenodo.org)... 137.138.52.235, 188.185.43.153, 188.185.48.75, ...\n",
            "Connecting to zenodo.org (zenodo.org)|137.138.52.235|:443... connected.\n",
            "HTTP request sent, awaiting response... 301 MOVED PERMANENTLY\n",
            "Location: /records/582600/files/mutant_R2.fastq [following]\n",
            "--2026-01-06 05:22:43--  https://zenodo.org/records/582600/files/mutant_R2.fastq\n",
            "Reusing existing connection to zenodo.org:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 4137809 (3.9M) [application/octet-stream]\n",
            "Saving to: ‘mutant_R2.fastq’\n",
            "\n",
            "mutant_R2.fastq     100%[===================>]   3.95M  1.05MB/s    in 3.7s    \n",
            "\n",
            "2026-01-06 05:22:48 (1.05 MB/s) - ‘mutant_R2.fastq’ saved [4137809/4137809]\n",
            "\n",
            "Started analysis of mutant_R1.fastq\n",
            "Approx 5% complete for mutant_R1.fastq\n",
            "Approx 15% complete for mutant_R1.fastq\n",
            "Approx 20% complete for mutant_R1.fastq\n",
            "Approx 30% complete for mutant_R1.fastq\n",
            "Approx 40% complete for mutant_R1.fastq\n",
            "Approx 45% complete for mutant_R1.fastq\n",
            "Approx 55% complete for mutant_R1.fastq\n",
            "Approx 60% complete for mutant_R1.fastq\n",
            "Approx 70% complete for mutant_R1.fastq\n",
            "Approx 80% complete for mutant_R1.fastq\n",
            "Approx 85% complete for mutant_R1.fastq\n",
            "Approx 95% complete for mutant_R1.fastq\n",
            "Analysis complete for mutant_R1.fastq\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_0352a0aa-4340-4ee4-9f96-6bd2f089631a\", \"mutant_R1_fastqc.html\", 543669)"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Started analysis of mutant_R2.fastq\n",
            "Approx 5% complete for mutant_R2.fastq\n",
            "Approx 15% complete for mutant_R2.fastq\n",
            "Approx 20% complete for mutant_R2.fastq\n",
            "Approx 30% complete for mutant_R2.fastq\n",
            "Approx 40% complete for mutant_R2.fastq\n",
            "Approx 45% complete for mutant_R2.fastq\n",
            "Approx 55% complete for mutant_R2.fastq\n",
            "Approx 60% complete for mutant_R2.fastq\n",
            "Approx 70% complete for mutant_R2.fastq\n",
            "Approx 80% complete for mutant_R2.fastq\n",
            "Approx 85% complete for mutant_R2.fastq\n",
            "Approx 95% complete for mutant_R2.fastq\n",
            "Analysis complete for mutant_R2.fastq\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_f429909f-219d-4f19-9866-0389a1010125\", \"mutant_R2_fastqc.html\", 547109)"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[0.000001] Reading FastQ file /content/mutant_R1.fastq;\n",
            "[0.055167] 12480 sequences found\n",
            "[0.055186] Done\n",
            "[0.055227] Reading FastQ file /content/mutant_R2.fastq;\n",
            "[0.111712] 12480 sequences found\n",
            "[0.111728] Done\n",
            "[0.111789] Reading read set file ASM_NAME/Sequences;\n",
            "[0.116512] 24960 sequences found\n",
            "[0.139931] Done\n",
            "[0.139954] 24960 sequences in total.\n",
            "[0.140054] Writing into roadmap file ASM_NAME/Roadmaps...\n",
            "[0.185702] Inputting sequences...\n",
            "[0.185730] Inputting sequence 0 / 24960\n",
            "[1.036538]  === Sequences loaded in 0.850838 s\n",
            "[1.036579] Done inputting sequences\n",
            "[1.036582] Destroying splay table\n",
            "[1.047863] Splay table destroyed\n",
            "[0.000000] Reading roadmap file ASM_NAME/Roadmaps\n",
            "[0.050627] 24960 roadmaps read\n",
            "[0.050848] Creating insertion markers\n",
            "[0.054673] Ordering insertion markers\n",
            "[0.075062] Counting preNodes\n",
            "[0.079014] 64223 preNodes counted, creating them now\n",
            "[0.191719] Adjusting marker info...\n",
            "[0.195632] Connecting preNodes\n",
            "[0.232364] Cleaning up memory\n",
            "[0.232512] Done creating preGraph\n",
            "[0.232520] Concatenation...\n",
            "[0.254865] Renumbering preNodes\n",
            "[0.254908] Initial preNode count 64223\n",
            "[0.257416] Destroyed 37924 preNodes\n",
            "[0.257437] Concatenation over!\n",
            "[0.257440] Clipping short tips off preGraph\n",
            "[0.258932] Concatenation...\n",
            "[0.264059] Renumbering preNodes\n",
            "[0.264084] Initial preNode count 26299\n",
            "[0.265847] Destroyed 3832 preNodes\n",
            "[0.265868] Concatenation over!\n",
            "[0.265870] 1986 tips cut off\n",
            "[0.265872] 22467 nodes left\n",
            "[0.265990] Writing into pregraph file ASM_NAME/PreGraph...\n",
            "[0.341775] Reading read set file ASM_NAME/Sequences;\n",
            "[0.346473] 24960 sequences found\n",
            "[0.370725] Done\n",
            "[0.423926] Reading pre-graph file ASM_NAME/PreGraph\n",
            "[0.424022] Graph has 22467 nodes and 24960 sequences\n",
            "[0.452301] Scanning pre-graph file ASM_NAME/PreGraph for k-mers\n",
            "[0.457704] 434645 kmers found\n",
            "[0.488047] Sorting kmer occurrence table ... \n",
            "[0.614960] Sorting done.\n",
            "[0.614990] Computing acceleration table... \n",
            "[0.658478] Computing offsets... \n",
            "[0.661007] Ghost Threading through reads 0 / 24960\n",
            "[0.661324]  === Ghost-Threaded in 0.000317 s\n",
            "[0.661333] Threading through reads 0 / 24960\n",
            "[1.748702]  === Threaded in 1.087368 s\n",
            "[1.753741] Correcting graph with cutoff 0.200000\n",
            "[1.755366] Determining eligible starting points\n",
            "[1.777196] Done listing starting nodes\n",
            "[1.777230] Initializing todo lists\n",
            "[1.781755] Done with initialization\n",
            "[1.781789] Activating arc lookup table\n",
            "[1.784236] Done activating arc lookup table\n",
            "[1.901149] 10000 / 22467 nodes visited\n",
            "[2.020593] 20000 / 22467 nodes visited\n",
            "[2.046823] Concatenation...\n",
            "[2.047028] Renumbering nodes\n",
            "[2.047037] Initial node count 22467\n",
            "[2.047106] Removed 22097 null nodes\n",
            "[2.047112] Concatenation over!\n",
            "[2.047114] Clipping short tips off graph, drastic\n",
            "[2.047142] Concatenation...\n",
            "[2.047313] Renumbering nodes\n",
            "[2.047318] Initial node count 370\n",
            "[2.047329] Removed 14 null nodes\n",
            "[2.047331] Concatenation over!\n",
            "[2.047335] 356 nodes left\n",
            "[2.047592] Writing into graph file ASM_NAME/Graph...\n",
            "[2.071303] WARNING: NO COVERAGE CUTOFF PROVIDED\n",
            "[2.071328] Velvet will probably leave behind many detectable errors\n",
            "[2.071331] See manual for instructions on how to set the coverage cutoff parameter\n",
            "[2.071334] Removing contigs with coverage < -1.000000...\n",
            "[2.071360] Concatenation...\n",
            "[2.071382] Renumbering nodes\n",
            "[2.071384] Initial node count 356\n",
            "[2.071387] Removed 0 null nodes\n",
            "[2.071389] Concatenation over!\n",
            "[2.071408] Concatenation...\n",
            "[2.071426] Renumbering nodes\n",
            "[2.071431] Initial node count 356\n",
            "[2.071436] Removed 0 null nodes\n",
            "[2.071438] Concatenation over!\n",
            "[2.071440] Clipping short tips off graph, drastic\n",
            "[2.071450] Concatenation...\n",
            "[2.071466] Renumbering nodes\n",
            "[2.071471] Initial node count 356\n",
            "[2.071476] Removed 0 null nodes\n",
            "[2.071477] Concatenation over!\n",
            "[2.071479] 356 nodes left\n",
            "[2.071481] WARNING: NO EXPECTED COVERAGE PROVIDED\n",
            "[2.071485] Velvet will be unable to resolve any repeats\n",
            "[2.071487] See manual for instructions on how to set the expected coverage parameter\n",
            "[2.071489] Concatenation...\n",
            "[2.071504] Renumbering nodes\n",
            "[2.071510] Initial node count 356\n",
            "[2.071513] Removed 0 null nodes\n",
            "[2.071519] Concatenation over!\n",
            "[2.071524] Removing reference contigs with coverage < -1.000000...\n",
            "[2.071545] Concatenation...\n",
            "[2.071563] Renumbering nodes\n",
            "[2.071568] Initial node count 356\n",
            "[2.071694] Removed 0 null nodes\n",
            "[2.071705] Concatenation over!\n",
            "[2.071936] Writing contigs into ASM_NAME/contigs.fa...\n",
            "[2.086812] Writing into stats file ASM_NAME/stats.txt...\n",
            "[2.087929] Writing into graph file ASM_NAME/LastGraph...\n",
            "Final graph has 356 nodes and n50 of 5029, max 12902, total 184065, using 0/24960 reads\n"
          ]
        }
      ],
      "source": [
        "!apt-get install fastqc\n",
        "!apt-get install velvet\n",
        "!wget 'https://zenodo.org/record/582600/files/mutant_R1.fastq'\n",
        "!wget 'https://zenodo.org/record/582600/files/mutant_R2.fastq'\n",
        "import os\n",
        "from google.colab import files\n",
        "get_ipython().system(\"fastqc mutant_R1.fastq\")\n",
        "files.download('mutant_R1_fastqc.html')\n",
        "get_ipython().system(\"fastqc mutant_R2.fastq\")\n",
        "files.download('mutant_R2_fastqc.html')\n",
        "!mkdir assembly_module\n",
        "os.chdir('assembly_module')\n",
        "!mkdir data\n",
        "get_ipython().system(\"velveth ASM_NAME 31 -short -separate -fastq /content/mutant_R1.fastq /content/mutant_R2.fastq\")\n",
        "get_ipython().system(\"velvetg ASM_NAME\")"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPU9+Qs7ya1duzSAaQWyIze",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}