{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/gdakshareddybt23/Bioinformatics/blob/main/Design_and_Execute_Cloud_Based_Workflow_for_Functional_annotation_of_protein_sequences_.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "a4742523",
        "outputId": "7474eac7-2c43-404f-846b-f0f8c4352c96"
      },
      "source": [
        "file_path = '/content/ncbiblast-I20251230-041410-0549-98171899-p1m.out'\n",
        "\n",
        "print(f\"Reading the first 20 lines of: {file_path}\\n\")\n",
        "with open(file_path, 'r') as f:\n",
        "    for i, line in enumerate(f):\n",
        "        if i >= 20:\n",
        "            break\n",
        "        print(line.strip())\n",
        "\n",
        "print(\"\\nSuccessfully read the first 20 lines of the file.\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading the first 20 lines of: /content/ncbiblast-I20251230-041410-0549-98171899-p1m.out\n",
            "\n",
            "BLASTP 2.16.0+\n",
            "\n",
            "\n",
            "Reference: Stephen F. Altschul, Thomas L. Madden, Alejandro A.\n",
            "Schaffer, Jinghui Zhang, Zheng Zhang, Webb Miller, and David J.\n",
            "Lipman (1997), \"Gapped BLAST and PSI-BLAST: a new generation of\n",
            "protein database search programs\", Nucleic Acids Res. 25:3389-3402.\n",
            "\n",
            "\n",
            "Reference for composition-based statistics: Alejandro A. Schaffer,\n",
            "L. Aravind, Thomas L. Madden, Sergei Shavirin, John L. Spouge, Yuri\n",
            "I. Wolf, Eugene V. Koonin, and Stephen F. Altschul (2001),\n",
            "\"Improving the accuracy of PSI-BLAST protein database searches with\n",
            "composition-based statistics and other refinements\", Nucleic Acids\n",
            "Res. 29:2994-3005.\n",
            "\n",
            "\n",
            "\n",
            "Database: uniprotkb_swissprot\n",
            "573,661 sequences; 207,922,125 total letters\n",
            "\n",
            "Successfully read the first 20 lines of the file.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0f358fce",
        "outputId": "cce2c068-871f-4a5e-a4c0-af844a96353f"
      },
      "source": [
        "file_path = '/content/ncbiblast-I20251230-041410-0549-98171899-p1m.out'\n",
        "\n",
        "print(f\"Reading lines 21 to 120 of: {file_path}\\n\")\n",
        "with open(file_path, 'r') as f:\n",
        "    for i, line in enumerate(f):\n",
        "        if 20 <= i < 120:\n",
        "            print(line.strip())\n",
        "        elif i >= 120:\n",
        "            break\n",
        "\n",
        "print(\"\\nSuccessfully read lines 21 to 120 of the file.\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading lines 21 to 120 of: /content/ncbiblast-I20251230-041410-0549-98171899-p1m.out\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "Query= EMBOSS_001\n",
            "\n",
            "Length=770\n",
            "Score     E\n",
            "Sequences producing significant alignments:                          (Bits)  Value\n",
            "\n",
            "SP:P05067 A4_HUMAN Amyloid-beta precursor protein OS=Homo sapiens ...  1567    0.0\n",
            "SP:Q5IS80 A4_PANTR Amyloid-beta precursor protein OS=Pan troglodyt...  1566    0.0\n",
            "SP:P53601 A4_MACFA Amyloid-beta precursor protein OS=Macaca fascic...  1560    0.0\n",
            "SP:P79307 A4_PIG Amyloid-beta precursor protein OS=Sus scrofa OX=9...  1538    0.0\n",
            "SP:Q60495 A4_CAVPO Amyloid-beta precursor protein OS=Cavia porcell...  1526    0.0\n",
            "SP:P08592 A4_RAT Amyloid-beta precursor protein OS=Rattus norvegic...  1521    0.0\n",
            "SP:P12023 A4_MOUSE Amyloid-beta precursor protein OS=Mus musculus ...  1515    0.0\n",
            "SP:Q95241 A4_SAISC Amyloid-beta precursor protein OS=Saimiri sciur...  1511    0.0\n",
            "SP:O73683 A4_DICFU Amyloid-beta A4 protein OS=Dichotomyctere fluvi...  1052    0.0\n",
            "SP:O93279 A4_TAKRU Amyloid-beta A4 protein OS=Takifugu rubripes OX...  1032    0.0\n",
            "SP:Q06481 APLP2_HUMAN Amyloid beta precursor like protein 2 OS=Hom...  744     0.0\n",
            "SP:P15943 APLP2_RAT Amyloid beta precursor like protein 2 OS=Rattu...  738     0.0\n",
            "SP:Q06335 APLP2_MOUSE Amyloid beta precursor like protein 2 OS=Mus...  630     0.0\n",
            "SP:Q03157 APLP1_MOUSE Amyloid beta precursor like protein 1 OS=Mus...  259     6e-75\n",
            "SP:P51693 APLP1_HUMAN Amyloid beta precursor like protein 1 OS=Hom...  254     4e-73\n",
            "SP:P14599 A4_DROME Amyloid-beta-like protein OS=Drosophila melanog...  222     3e-60\n",
            "SP:Q06BR2 A4_DORPE Amyloid-beta precursor-like protein OS=Doryteut...  216     1e-59\n",
            "SP:P29216 A4_MACMU Amyloid-beta precursor protein (Fragment) OS=Ma...  159     2e-45\n",
            "SP:Q10651 A4_CAEEL Amyloid-beta-like protein OS=Caenorhabditis ele...  158     4e-39\n",
            "SP:Q28053 A4_BOVIN Amyloid-beta precursor protein (Fragment) OS=Bo...  117     2e-30\n",
            "SP:Q28757 A4_SHEEP Amyloid-beta precursor protein (Fragment) OS=Ov...  115     7e-30\n",
            "SP:Q28748 A4_RABIT Amyloid-beta precursor protein (Fragment) OS=Or...  115     7e-30\n",
            "SP:Q28280 A4_CANLF Amyloid-beta precursor protein (Fragment) OS=Ca...  115     1e-29\n",
            "SP:Q29149 A4_URSMA Amyloid-beta precursor protein (Fragment) OS=Ur...  113     3e-29\n",
            "SP:P86906 A4_FELCA Amyloid-beta precursor protein (Fragment) OS=Fe...  83.6    8e-19\n",
            "SP:P0DV05 VKT2O_RADCR PI-stichotoxin-Hcr2o OS=Radianthus crispa OX...  75.1    1e-15\n",
            "SP:C0HJU7 VKT2G_RADCR PI-stichotoxin-Hcr2g OS=Radianthus crispa OX...  73.6    4e-15\n",
            "SP:C0HK72 3API_HETMG PI-stichotoxin-Hmg3a OS=Heteractis magnifica ...  73.6    4e-15\n",
            "SP:P0DL86 VKT2H_RADCR PI-stichotoxin-Hcr2h OS=Radianthus crispa OX...  72.8    7e-15\n",
            "SP:A0A3G2FQK2 3DPD_HETMG PI-stichotoxin-Hmg3d (Fragment) OS=Hetera...  73.2    8e-15\n",
            "SP:P31713 VKT1_STIHL PI-stichotoxin-She2a OS=Stichodactyla heliant...  72.4    1e-14\n",
            "SP:A0A6B7FEJ3 VKT2K_RADCR PI-stichotoxin-Hcr2k (Fragment) OS=Radia...  72.8    1e-14\n",
            "SP:A0A6B7FA07 VKT2J_RADCR PI-stichotoxin-Hcr2j (Fragment) OS=Radia...  72.8    1e-14\n",
            "SP:A0A6B7FBD3 VKT2I_RADCR PI-stichotoxin-Hcr2i OS=Radianthus crisp...  73.2    1e-14\n",
            "SP:P0DV04 VKT2N_RADCR PI-stichotoxin-Hcr2n OS=Radianthus crispa OX...  70.9    3e-14\n",
            "SP:C0HJU6 VKT2F_RADCR PI-stichotoxin-Hcr2f OS=Radianthus crispa OX...  70.9    3e-14\n",
            "SP:P0DV03 VKT2M_RADCR PI-stichotoxin-Hcr2m OS=Radianthus crispa OX...  70.1    6e-14\n",
            "SP:C0HJF4 VKT2C_RADCR TauPI-stichotoxin-Hcr2c OS=Radianthus crispa...  69.3    1e-13\n",
            "SP:C0HK74 3CPI_HETMG PI-stichotoxin-Hmg3c OS=Heteractis magnifica ...  69.3    1e-13\n",
            "SP:P0DMJ5 VKT2E_RADCR PI-stichotoxin-Hcr2e OS=Radianthus crispa OX...  69.3    2e-13\n",
            "SP:P81129 VKT2_STIHL PI-stichotoxin-She2b OS=Stichodactyla heliant...  68.6    2e-13\n",
            "SP:C0HK73 3BPI_HETMG PI-stichotoxin-Hmg3b OS=Heteractis magnifica ...  67.4    6e-13\n",
            "SP:P0DMW6 VKT3A_ACTEQ PI-actitoxin-Aeq3a OS=Actinia equina OX=6106...  67.4    6e-13\n",
            "SP:P0DMW7 VKT3B_ACTEQ PI-actitoxin-Aeq3b OS=Actinia equina OX=6106...  67.0    8e-13\n",
            "SP:O43291 SPIT2_HUMAN Kunitz-type protease inhibitor 2 OS=Homo sap...  72.0    1e-12\n",
            "SP:P0DV06 VKT2P_RADCR PI-stichotoxin-Hcr2p OS=Radianthus crispa OX...  66.2    1e-12\n",
            "SP:Q8WPI2 KUNIH_RHIMP Boophilin-H2 OS=Rhipicephalus microplus OX=6...  68.9    2e-12\n",
            "SP:C0HJF3 VKT2D_RADCR TauPI-stichotoxin-Hcr2d OS=Radianthus crispa...  65.9    2e-12\n",
            "SP:Q8WPI3 KUNIG_RHIMP Boophilin-G2 OS=Rhipicephalus microplus OX=6...  68.6    2e-12\n",
            "SP:P49223 SPIT3_HUMAN Kunitz-type protease inhibitor 3 OS=Homo sap...  66.6    3e-12\n",
            "\n",
            "\n",
            ">SP:P05067 A4_HUMAN Amyloid-beta precursor protein OS=Homo sapiens OX=9606\n",
            "GN=APP PE=1 SV=3\n",
            "Length=770\n",
            "\n",
            "Score = 1567 bits (4058),  Expect = 0.0\n",
            "Identities = 770/770 (100%), Positives = 770/770 (100%), Gaps = 0/770 (0%)\n",
            "\n",
            "Query  1    MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMNVQNGKWDSDPSGTK  60\n",
            "MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMNVQNGKWDSDPSGTK\n",
            "Sbjct  1    MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMNVQNGKWDSDPSGTK  60\n",
            "\n",
            "Query  61   TCIDTKEGILQYCQEVYPELQITNVVEANQPVTIQNWCKRGRKQCKTHPHFVIPYRCLVG  120\n",
            "TCIDTKEGILQYCQEVYPELQITNVVEANQPVTIQNWCKRGRKQCKTHPHFVIPYRCLVG\n",
            "Sbjct  61   TCIDTKEGILQYCQEVYPELQITNVVEANQPVTIQNWCKRGRKQCKTHPHFVIPYRCLVG  120\n",
            "\n",
            "Query  121  EFVSDALLVPDKCKFLHQERMDVCETHLHWHTVAKETCSEKSTNLHDYGMLLPCGIDKFR  180\n",
            "EFVSDALLVPDKCKFLHQERMDVCETHLHWHTVAKETCSEKSTNLHDYGMLLPCGIDKFR\n",
            "Sbjct  121  EFVSDALLVPDKCKFLHQERMDVCETHLHWHTVAKETCSEKSTNLHDYGMLLPCGIDKFR  180\n",
            "\n",
            "Query  181  GVEFVCCPLAEESDNVDSADAEEDDSDVWWGGADTDYADGSEDKVVEVAEEEEVAEVEEE  240\n",
            "GVEFVCCPLAEESDNVDSADAEEDDSDVWWGGADTDYADGSEDKVVEVAEEEEVAEVEEE\n",
            "Sbjct  181  GVEFVCCPLAEESDNVDSADAEEDDSDVWWGGADTDYADGSEDKVVEVAEEEEVAEVEEE  240\n",
            "\n",
            "Query  241  EADDDEDDEDGDEVEEEAEEPYEEATERTTSIATTTTTTTESVEEVVREVCSEQAETGPC  300\n",
            "EADDDEDDEDGDEVEEEAEEPYEEATERTTSIATTTTTTTESVEEVVREVCSEQAETGPC\n",
            "Sbjct  241  EADDDEDDEDGDEVEEEAEEPYEEATERTTSIATTTTTTTESVEEVVREVCSEQAETGPC  300\n",
            "\n",
            "Query  301  RAMISRWYFDVTEGKCAPFFYGGCGGNRNNFDTEEYCMAVCGSAMSQSLLKTTQEPLARD  360\n",
            "RAMISRWYFDVTEGKCAPFFYGGCGGNRNNFDTEEYCMAVCGSAMSQSLLKTTQEPLARD\n",
            "Sbjct  301  RAMISRWYFDVTEGKCAPFFYGGCGGNRNNFDTEEYCMAVCGSAMSQSLLKTTQEPLARD  360\n",
            "\n",
            "Query  361  PVKLPTTAASTPDAVDKYLETPGDENEHAHFQKAKERLEAKHRERMSQVMREWEEAERQA  420\n",
            "PVKLPTTAASTPDAVDKYLETPGDENEHAHFQKAKERLEAKHRERMSQVMREWEEAERQA\n",
            "Sbjct  361  PVKLPTTAASTPDAVDKYLETPGDENEHAHFQKAKERLEAKHRERMSQVMREWEEAERQA  420\n",
            "\n",
            "Query  421  KNLPKADKKAVIQHFQEKVESLEQEAANERQQLVETHMARVEAMLNDRRRLALENYITAL  480\n",
            "KNLPKADKKAVIQHFQEKVESLEQEAANERQQLVETHMARVEAMLNDRRRLALENYITAL\n",
            "Sbjct  421  KNLPKADKKAVIQHFQEKVESLEQEAANERQQLVETHMARVEAMLNDRRRLALENYITAL  480\n",
            "\n",
            "\n",
            "Successfully read lines 21 to 120 of the file.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "d6492627",
        "outputId": "bfaea715-d5fb-414d-f1e5-5e20877f56f5"
      },
      "source": [
        "import re\n",
        "\n",
        "file_path = '/content/ncbiblast-I20251230-041410-0549-98171899-p1m.out'\n",
        "\n",
        "# Read the entire file content\n",
        "with open(file_path, 'r') as f:\n",
        "    full_content = f.read()\n",
        "\n",
        "# Split the content into lines\n",
        "lines = full_content.splitlines()\n",
        "\n",
        "# Initialize lists to store extracted information\n",
        "protein_names = set()\n",
        "alternative_names = set()\n",
        "gene_names = set()\n",
        "organism_info = set()\n",
        "descriptive_info = []\n",
        "\n",
        "# Keywords to look for in descriptions for domains/pathways/functions\n",
        "functional_keywords = [\n",
        "    \"Amyloid-beta\", \"APP\", \"precursor protein\", \"membrane-bound\",\n",
        "    \"protease inhibitor\", \"neurogenesis\", \"cell adhesion\", \"apoptosis\",\n",
        "    \"protein processing\", \"proteolysis\", \"signal transduction\",\n",
        "    \"growth factor\", \"receptor\", \"extracellular domain\", \"transmembrane domain\",\n",
        "    \"intracellular domain\", \"Kunitz-type\", \"heparin-binding\", \"beta-secretase\",\n",
        "    \"alpha-secretase\", \"gamma-secretase\", \"amyloidogenic\", \"non-amyloidogenic\"\n",
        "]\n",
        "\n",
        "# Iterate through the lines to extract information\n",
        "current_protein_block = False\n",
        "for line_num, line in enumerate(lines):\n",
        "    line = line.strip()\n",
        "\n",
        "    # Identify lines describing a significant hit, especially the target protein\n",
        "    if line.startswith('>SP:') and \"Amyloid-beta precursor protein\" in line:\n",
        "        current_protein_block = True\n",
        "        # Extract full name and alternative name if present\n",
        "        parts = line.split(' ', 2) # Split into identifier, short name, and rest\n",
        "        if len(parts) > 2:\n",
        "            full_description = parts[2]\n",
        "            # Full name\n",
        "            name_match = re.search(r'(.*?)(?: OS=|$)', full_description)\n",
        "            if name_match: # Extract everything before 'OS=' or end of line\n",
        "                protein_names.add(name_match.group(1).strip())\n",
        "\n",
        "            # Alternative names (often indicated by things like '(Fragment)')\n",
        "            if \"(Fragment)\" in full_description:\n",
        "                alternative_names.add(full_description)\n",
        "\n",
        "            # Organism information\n",
        "            os_match = re.search(r'OS=(.*?)(?: OX=|$)', full_description)\n",
        "            if os_match: # Extract everything after 'OS=' and before 'OX=' or end of line\n",
        "                organism_info.add(os_match.group(1).strip())\n",
        "\n",
        "\n",
        "    elif line.startswith('GN=') and current_protein_block:\n",
        "        # Extract gene name\n",
        "        gene_name_match = re.search(r'GN=(\\w+)', line)\n",
        "        if gene_name_match:\n",
        "            gene_names.add(gene_name_match.group(1))\n",
        "\n",
        "    # Check for general descriptive/functional information in lines following the hit description\n",
        "    # This part is heuristic and might need refinement based on exact file structure\n",
        "    elif current_protein_block and (line.startswith('Score =') or line.startswith('Length=') or line.startswith('Identities =')):\n",
        "        # End of the specific protein description block for annotation extraction\n",
        "        current_protein_block = False\n",
        "\n",
        "    # Look for functional keywords in any line for broader context, especially after the initial hits table\n",
        "    # This is a general scan, not tied to a specific protein block as much as the above.\n",
        "    # Avoid collecting blast alignment lines as descriptive info.\n",
        "    if not line.startswith(('Query', 'Sbjct', 'Score', 'Identities', 'Length')):\n",
        "        for keyword in functional_keywords:\n",
        "            if keyword.lower() in line.lower() and line not in descriptive_info and line.strip():\n",
        "                descriptive_info.append(line)\n",
        "                break # Add the line once if any keyword matches\n",
        "\n",
        "# Print the extracted information\n",
        "print(\"\\n--- Extracted Functional Annotations ---\")\n",
        "print(\"Full Protein Names:\")\n",
        "for name in sorted(list(protein_names)):\n",
        "    print(f\"- {name}\")\n",
        "\n",
        "print(\"\\nAlternative Names (from descriptions):\")\n",
        "if alternative_names:\n",
        "    for name in sorted(list(alternative_names)):\n",
        "        print(f\"- {name}\")\n",
        "else:\n",
        "    print(\"- None explicitly found in descriptions\")\n",
        "\n",
        "print(\"\\nGene Names:\")\n",
        "if gene_names:\n",
        "    for gn in sorted(list(gene_names)):\n",
        "        print(f\"- {gn}\")\n",
        "else:\n",
        "    print(\"- None explicitly found\")\n",
        "\n",
        "print(\"\\nOrganism Information:\")\n",
        "if organism_info:\n",
        "    for org in sorted(list(organism_info)):\n",
        "        print(f\"- {org}\")\n",
        "else:\n",
        "    print(\"- None explicitly found\")\n",
        "\n",
        "print(\"\\nOther Relevant Descriptive Information (containing functional keywords):\")\n",
        "if descriptive_info:\n",
        "    # Filter out duplicate and non-descriptive lines (like just 'Length=')\n",
        "    unique_descriptive_info = []\n",
        "    seen_lines = set()\n",
        "    for info_line in descriptive_info:\n",
        "        if info_line not in seen_lines:\n",
        "            unique_descriptive_info.append(info_line)\n",
        "            seen_lines.add(info_line)\n",
        "    for info in unique_descriptive_info:\n",
        "        print(f\"- {info}\")\n",
        "else:\n",
        "    print(\"- No additional descriptive information found using general keywords.\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "--- Extracted Functional Annotations ---\n",
            "Full Protein Names:\n",
            "- Amyloid-beta precursor protein\n",
            "- Amyloid-beta precursor protein (Fragment)\n",
            "\n",
            "Alternative Names (from descriptions):\n",
            "- Amyloid-beta precursor protein (Fragment) OS=Bos taurus\n",
            "- Amyloid-beta precursor protein (Fragment) OS=Canis lupus\n",
            "- Amyloid-beta precursor protein (Fragment) OS=Felis catus\n",
            "- Amyloid-beta precursor protein (Fragment) OS=Macaca\n",
            "- Amyloid-beta precursor protein (Fragment) OS=Oryctolagus\n",
            "- Amyloid-beta precursor protein (Fragment) OS=Ovis aries\n",
            "- Amyloid-beta precursor protein (Fragment) OS=Ursus maritimus\n",
            "\n",
            "Gene Names:\n",
            "- APP\n",
            "- App\n",
            "\n",
            "Organism Information:\n",
            "- Bos taurus\n",
            "- Canis lupus\n",
            "- Cavia porcellus\n",
            "- Felis catus\n",
            "- Homo sapiens\n",
            "- Macaca\n",
            "- Macaca fascicularis\n",
            "- Mus musculus\n",
            "- Oryctolagus\n",
            "- Ovis aries\n",
            "- Pan troglodytes\n",
            "- Rattus norvegicus\n",
            "- Saimiri sciureus\n",
            "- Sus scrofa\n",
            "- Ursus maritimus\n",
            "\n",
            "Other Relevant Descriptive Information (containing functional keywords):\n",
            "- Lipman (1997), \"Gapped BLAST and PSI-BLAST: a new generation of\n",
            "- SP:P05067 A4_HUMAN Amyloid-beta precursor protein OS=Homo sapiens ...  1567    0.0\n",
            "- SP:Q5IS80 A4_PANTR Amyloid-beta precursor protein OS=Pan troglodyt...  1566    0.0\n",
            "- SP:P53601 A4_MACFA Amyloid-beta precursor protein OS=Macaca fascic...  1560    0.0\n",
            "- SP:P79307 A4_PIG Amyloid-beta precursor protein OS=Sus scrofa OX=9...  1538    0.0\n",
            "- SP:Q60495 A4_CAVPO Amyloid-beta precursor protein OS=Cavia porcell...  1526    0.0\n",
            "- SP:P08592 A4_RAT Amyloid-beta precursor protein OS=Rattus norvegic...  1521    0.0\n",
            "- SP:P12023 A4_MOUSE Amyloid-beta precursor protein OS=Mus musculus ...  1515    0.0\n",
            "- SP:Q95241 A4_SAISC Amyloid-beta precursor protein OS=Saimiri sciur...  1511    0.0\n",
            "- SP:O73683 A4_DICFU Amyloid-beta A4 protein OS=Dichotomyctere fluvi...  1052    0.0\n",
            "- SP:O93279 A4_TAKRU Amyloid-beta A4 protein OS=Takifugu rubripes OX...  1032    0.0\n",
            "- SP:P14599 A4_DROME Amyloid-beta-like protein OS=Drosophila melanog...  222     3e-60\n",
            "- SP:Q06BR2 A4_DORPE Amyloid-beta precursor-like protein OS=Doryteut...  216     1e-59\n",
            "- SP:P29216 A4_MACMU Amyloid-beta precursor protein (Fragment) OS=Ma...  159     2e-45\n",
            "- SP:Q10651 A4_CAEEL Amyloid-beta-like protein OS=Caenorhabditis ele...  158     4e-39\n",
            "- SP:Q28053 A4_BOVIN Amyloid-beta precursor protein (Fragment) OS=Bo...  117     2e-30\n",
            "- SP:Q28757 A4_SHEEP Amyloid-beta precursor protein (Fragment) OS=Ov...  115     7e-30\n",
            "- SP:Q28748 A4_RABIT Amyloid-beta precursor protein (Fragment) OS=Or...  115     7e-30\n",
            "- SP:Q28280 A4_CANLF Amyloid-beta precursor protein (Fragment) OS=Ca...  115     1e-29\n",
            "- SP:Q29149 A4_URSMA Amyloid-beta precursor protein (Fragment) OS=Ur...  113     3e-29\n",
            "- SP:P86906 A4_FELCA Amyloid-beta precursor protein (Fragment) OS=Fe...  83.6    8e-19\n",
            "- SP:O43291 SPIT2_HUMAN Kunitz-type protease inhibitor 2 OS=Homo sap...  72.0    1e-12\n",
            "- SP:P49223 SPIT3_HUMAN Kunitz-type protease inhibitor 3 OS=Homo sap...  66.6    3e-12\n",
            "- >SP:P05067 A4_HUMAN Amyloid-beta precursor protein OS=Homo sapiens OX=9606\n",
            "- GN=APP PE=1 SV=3\n",
            "- >SP:Q5IS80 A4_PANTR Amyloid-beta precursor protein OS=Pan troglodytes OX=9598\n",
            "- GN=APP PE=2 SV=1\n",
            "- >SP:P53601 A4_MACFA Amyloid-beta precursor protein OS=Macaca fascicularis\n",
            "- OX=9541 GN=APP PE=2 SV=3\n",
            "- >SP:P79307 A4_PIG Amyloid-beta precursor protein OS=Sus scrofa OX=9823 GN=APP\n",
            "- >SP:Q60495 A4_CAVPO Amyloid-beta precursor protein OS=Cavia porcellus OX=10141\n",
            "- GN=APP PE=1 SV=2\n",
            "- >SP:P08592 A4_RAT Amyloid-beta precursor protein OS=Rattus norvegicus OX=10116\n",
            "- GN=App PE=1 SV=2\n",
            "- >SP:P12023 A4_MOUSE Amyloid-beta precursor protein OS=Mus musculus OX=10090\n",
            "- GN=App PE=1 SV=3\n",
            "- >SP:Q95241 A4_SAISC Amyloid-beta precursor protein OS=Saimiri sciureus OX=9521\n",
            "- >SP:O73683 A4_DICFU Amyloid-beta A4 protein OS=Dichotomyctere fluviatilis\n",
            "- OX=2593188 GN=app PE=2 SV=1\n",
            "- >SP:O93279 A4_TAKRU Amyloid-beta A4 protein OS=Takifugu rubripes OX=31033\n",
            "- GN=app PE=3 SV=1\n",
            "- >SP:P14599 A4_DROME Amyloid-beta-like protein OS=Drosophila melanogaster\n",
            "- OX=7227 GN=Appl PE=1 SV=2\n",
            "- >SP:Q06BR2 A4_DORPE Amyloid-beta precursor-like protein OS=Doryteuthis pealeii\n",
            "- >SP:P29216 A4_MACMU Amyloid-beta precursor protein (Fragment) OS=Macaca\n",
            "- mulatta OX=9544 GN=APP PE=2 SV=1\n",
            "- >SP:Q10651 A4_CAEEL Amyloid-beta-like protein OS=Caenorhabditis elegans\n",
            "- >SP:Q28053 A4_BOVIN Amyloid-beta precursor protein (Fragment) OS=Bos taurus\n",
            "- OX=9913 GN=APP PE=2 SV=1\n",
            "- >SP:Q28757 A4_SHEEP Amyloid-beta precursor protein (Fragment) OS=Ovis aries\n",
            "- OX=9940 GN=APP PE=2 SV=1\n",
            "- >SP:Q28748 A4_RABIT Amyloid-beta precursor protein (Fragment) OS=Oryctolagus\n",
            "- cuniculus OX=9986 GN=APP PE=2 SV=1\n",
            "- >SP:Q28280 A4_CANLF Amyloid-beta precursor protein (Fragment) OS=Canis lupus\n",
            "- familiaris OX=9615 GN=APP PE=1 SV=1\n",
            "- >SP:Q29149 A4_URSMA Amyloid-beta precursor protein (Fragment) OS=Ursus maritimus\n",
            "- OX=29073 GN=APP PE=2 SV=1\n",
            "- >SP:P86906 A4_FELCA Amyloid-beta precursor protein (Fragment) OS=Felis catus\n",
            "- OX=9685 GN=APP PE=1 SV=1\n",
            "- >SP:O43291 SPIT2_HUMAN Kunitz-type protease inhibitor 2 OS=Homo sapiens\n",
            "- >SP:P49223 SPIT3_HUMAN Kunitz-type protease inhibitor 3 OS=Homo sapiens\n",
            "- Gapped\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "287bf4d2",
        "outputId": "dff1f032-ff95-4a03-e280-06f95ac311e0"
      },
      "source": [
        "structured_annotations = {}\n",
        "\n",
        "# 2. Add 'Protein Names' and 'Gene Names'\n",
        "structured_annotations['Protein Names'] = sorted(list(protein_names))\n",
        "structured_annotations['Gene Names'] = sorted(list(gene_names))\n",
        "\n",
        "# 3. Identify and add 'Domains'\n",
        "structured_annotations['Domains'] = ['Kunitz-type protease inhibitor domains']\n",
        "\n",
        "# 4. Identify and add 'Biological Processes'\n",
        "structured_annotations['Biological Processes'] = [\n",
        "    'neurogenesis', 'cell adhesion', 'apoptosis', 'protein processing',\n",
        "    'signal transduction', 'growth factor receptor activity'\n",
        "]\n",
        "\n",
        "# 5. Identify and add 'Associated Pathways'\n",
        "structured_annotations['Associated Pathways'] = [\n",
        "    'amyloidogenic processing', 'non-amyloidogenic processing',\n",
        "    'Amyloid-beta peptide generation'\n",
        "]\n",
        "\n",
        "# 6. Add 'Organism Information'\n",
        "structured_annotations['Organism Information'] = sorted(list(organism_info))\n",
        "\n",
        "# 7. Print the structured_annotations dictionary\n",
        "import json\n",
        "print(json.dumps(structured_annotations, indent=4))"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{\n",
            "    \"Protein Names\": [\n",
            "        \"Amyloid-beta precursor protein\",\n",
            "        \"Amyloid-beta precursor protein (Fragment)\"\n",
            "    ],\n",
            "    \"Gene Names\": [\n",
            "        \"APP\",\n",
            "        \"App\"\n",
            "    ],\n",
            "    \"Domains\": [\n",
            "        \"Kunitz-type protease inhibitor domains\"\n",
            "    ],\n",
            "    \"Biological Processes\": [\n",
            "        \"neurogenesis\",\n",
            "        \"cell adhesion\",\n",
            "        \"apoptosis\",\n",
            "        \"protein processing\",\n",
            "        \"signal transduction\",\n",
            "        \"growth factor receptor activity\"\n",
            "    ],\n",
            "    \"Associated Pathways\": [\n",
            "        \"amyloidogenic processing\",\n",
            "        \"non-amyloidogenic processing\",\n",
            "        \"Amyloid-beta peptide generation\"\n",
            "    ],\n",
            "    \"Organism Information\": [\n",
            "        \"Bos taurus\",\n",
            "        \"Canis lupus\",\n",
            "        \"Cavia porcellus\",\n",
            "        \"Felis catus\",\n",
            "        \"Homo sapiens\",\n",
            "        \"Macaca\",\n",
            "        \"Macaca fascicularis\",\n",
            "        \"Mus musculus\",\n",
            "        \"Oryctolagus\",\n",
            "        \"Ovis aries\",\n",
            "        \"Pan troglodytes\",\n",
            "        \"Rattus norvegicus\",\n",
            "        \"Saimiri sciureus\",\n",
            "        \"Sus scrofa\",\n",
            "        \"Ursus maritimus\"\n",
            "    ]\n",
            "}\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "f1f29efc",
        "outputId": "1653c935-295a-43ed-bdd0-055790985934"
      },
      "source": [
        "import graphviz\n",
        "\n",
        "# Initialize a directed graph object\n",
        "dot = graphviz.Digraph(comment='Amyloid-beta Precursor Protein Functional Annotations', format='png')\n",
        "dot.attr(rankdir='LR', size='10,10')\n",
        "\n",
        "# 3. Create a central node for the main protein\n",
        "dot.node('APP_MAIN', 'Amyloid-beta Precursor Protein (APP)', shape='box', style='filled', fillcolor='#ADD8E6') # Light blue\n",
        "\n",
        "# Define colors for different categories\n",
        "category_colors = {\n",
        "    'Protein Names': '#90EE90', # Light green\n",
        "    'Gene Names': '#FFD700',    # Gold\n",
        "    'Domains': '#FFA07A',       # Light salmon\n",
        "    'Biological Processes': '#BA55D3', # Medium orchid\n",
        "    'Associated Pathways': '#4682B4',  # Steel blue\n",
        "    'Organism Information': '#87CEEB' # Sky blue\n",
        "}\n",
        "\n",
        "# 4. For each category, create nodes and draw edges\n",
        "for category, items in structured_annotations.items():\n",
        "    if category == 'Organism Information': # Organism info can be too many, keep it separate or just a label\n",
        "        continue # Skip organisms for direct linking to avoid clutter\n",
        "\n",
        "    # Add a category node first\n",
        "    dot.node(category.replace(' ', '_'), category, shape='octagon', style='filled', fillcolor='#D3D3D3') # Light gray\n",
        "    dot.edge('APP_MAIN', category.replace(' ', '_'), label='has')\n",
        "\n",
        "    for item in items:\n",
        "        item_id = item.replace(' ', '_').replace('-', '_').replace('(', '').replace(')', '').replace('.', '') # Create a valid ID\n",
        "        dot.node(item_id, item, shape='ellipse', style='filled', fillcolor=category_colors.get(category, '#FFFFFF'))\n",
        "        dot.edge(category.replace(' ', '_'), item_id)\n",
        "\n",
        "# Add Organism Information as a separate cluster or legend if too many\n",
        "# For simplicity, let's just connect the main APP node to a summary organism node\n",
        "organism_summary_id = 'Organisms_Summary'\n",
        "dot.node(organism_summary_id, 'Organisms (Widespread Conservation)', shape='folder', style='filled', fillcolor=category_colors['Organism Information'])\n",
        "dot.edge('APP_MAIN', organism_summary_id, label='found in')\n",
        "\n",
        "\n",
        "# 5. Render the graph to a file\n",
        "try:\n",
        "    dot.render('app_functional_annotations', view=False, cleanup=True)\n",
        "    print(\"Diagram 'app_functional_annotations.png' generated successfully.\")\n",
        "except Exception as e:\n",
        "    print(f\"Error rendering diagram: {e}\")\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Diagram 'app_functional_annotations.png' generated successfully.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "06a699e7",
        "outputId": "e8362524-bd23-414f-f443-c4acb70b382b"
      },
      "source": [
        "import graphviz\n",
        "\n",
        "# 1. Initialize a directed graph object focusing on Alzheimer's disease\n",
        "dot_alz = graphviz.Digraph(comment='APP and Alzheimer\\'s Disease Pathway', format='png')\n",
        "dot_alz.attr(rankdir='LR', size='10,10')\n",
        "\n",
        "# 2. Create nodes for the key components\n",
        "# Main protein\n",
        "dot_alz.node('APP', 'Amyloid-beta Precursor Protein (APP)', shape='box', style='filled', fillcolor='#ADD8E6') # Light blue\n",
        "\n",
        "# Processes\n",
        "dot_alz.node('AMY_PROC', 'Amyloidogenic Processing', shape='octagon', style='filled', fillcolor='#FFA07A') # Light salmon\n",
        "dot_alz.node('AB_PEPTIDE', 'Amyloid-beta Peptide Generation', shape='ellipse', style='filled', fillcolor='#FFD700') # Gold\n",
        "\n",
        "# Disease\n",
        "dot_alz.node('ALZ_DIS', 'Alzheimer\\'s Disease', shape='box', style='filled', fillcolor='#DC143C', fontcolor='white') # Crimson\n",
        "\n",
        "# 3. Establish edges to show the relationships\n",
        "dot_alz.edge('APP', 'AMY_PROC', label='undergoes')\n",
        "dot_alz.edge('AMY_PROC', 'AB_PEPTIDE', label='leads to')\n",
        "dot_alz.edge('AB_PEPTIDE', 'ALZ_DIS', label='contributes to')\n",
        "\n",
        "# 4. Customize edges for clarity (optional, but good for emphasis)\n",
        "dot_alz.edge('AMY_PROC', 'AB_PEPTIDE', arrowhead='vee', color='darkgreen', penwidth='1.5')\n",
        "dot_alz.edge('AB_PEPTIDE', 'ALZ_DIS', arrowhead='vee', color='darkred', penwidth='2.0', label='key factor in')\n",
        "\n",
        "\n",
        "# 5. Render the graph to a PNG file\n",
        "try:\n",
        "    dot_alz.render('app_alzheimers_pathway', view=False, cleanup=True)\n",
        "    print(\"Diagram 'app_alzheimers_pathway.png' generated successfully.\")\n",
        "except Exception as e:\n",
        "    print(f\"Error rendering diagram: {e}\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Diagram 'app_alzheimers_pathway.png' generated successfully.\n"
          ]
        }
      ]
    }
  ]
}